atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStoreInvalidatorIntegrationTest.java

```diff
@@ -0,0 +1,155 @@
+            .with(new CassandraContainer());
+
+    private static final long ZERO = 0L;
+    private static final long ONE_MILLION = 1000000L;
+    private static final int POOL_SIZE = 16;
+    private static final int TIMEOUT_SECONDS = 5;
+
+    private final CassandraKeyValueService kv = CassandraKeyValueService.create(
+            CassandraKeyValueServiceConfigManager.createSimpleManager(CassandraContainer.KVS_CONFIG),
+            CassandraContainer.LEADER_CONFIG);
+    private final TimestampStoreInvalidator invalidator = new CassandraTimestampStoreInvalidator(kv);
+
+    private TimestampBoundStore timestampBoundStore;
+    private ExecutorService executorService;
+
+    @Before
+    public void setUp() {
+        kv.dropTable(AtlasDbConstants.TIMESTAMP_TABLE);
+        timestampBoundStore = CassandraTimestampBoundStore.create(kv);
+        executorService = Executors.newFixedThreadPool(POOL_SIZE);
+    }
+
+    @After
+    public void close() {
+        kv.dropTable(AtlasDbConstants.TIMESTAMP_TABLE);
+        kv.close();
+    }
+
+    @Test
+    public void canInvalidateTimestampTableIfItAlreadyExistsWithData() {
+        takeOutOneMillionTimestamps();
+        invalidateAndThenCheckCannotReadBound();
+    }
+
+    @Test
+    public void canInvalidateTimestampTableIfItAlreadyExistsWithoutData() {
+        invalidateAndThenCheckCannotReadBound();
+    }
+
+    @Test
+    public void canInvalidateTimestampTableIfItDoesNotExistYet() {
+        kv.dropTable(AtlasDbConstants.TIMESTAMP_TABLE);
+        invalidateAndThenCheckCannotReadBound();
+    }
+
+    @Test
+    public void invalidateIsIdempotent() {
+        invalidator.invalidateTimestampStore();
+        invalidateAndThenCheckCannotReadBound();
+    }
+
+    @Test
+    public void resilientToMultipleConcurrentInvalidations() {
+        takeOutOneMillionTimestamps();
+        executeInParallelOnExecutorService(this::invalidateAndThenCheckCannotReadBound);
+        revalidateAndThenCheckCanReadBound(ONE_MILLION);
+    }
+
+    private void takeOutOneMillionTimestamps() {
+        timestampBoundStore.getUpperLimit();
+        timestampBoundStore.storeUpperLimit(ONE_MILLION);
+    }
+
+    @Test
+    public void cannotGoBackInTimeWithRevalidation() {
+        invalidateAndThenCheckCannotReadBound();
+        revalidateAndThenCheckCanReadBound(ZERO);
+        takeOutOneMillionTimestamps();
+        revalidateAndThenCheckCanReadBound(ONE_MILLION); // in particular, not ZERO
+    }
+
+    @Test
+    public void canReadTimestampTableAfterRevalidation() {
+        takeOutOneMillionTimestamps();
+        invalidateAndThenCheckCannotReadBound();
+        revalidateAndThenCheckCanReadBound(ONE_MILLION);
+    }
+
+    @Test
+    public void resilientToMultipleConcurrentRevalidations() {
+        takeOutOneMillionTimestamps();
+        invalidateAndThenCheckCannotReadBound();
+        executeInParallelOnExecutorService(() -> revalidateAndThenCheckCanReadBound(ONE_MILLION));
+    }
+
```

Comment:
Thanks! I put the concurrency ones in as I was concerned about auto-migrations - in case multiple clients start up and all of them try to do these steps